src/vdaf/prio3.rs

```diff
@@ -39,396 +208,470 @@ pub struct InputShareMessage<F> {
-            blind: helper.blind,
+        // Prep the output messages.
+        let mut out = Vec::with_capacity(num_aggregators);
+        out.push(Prio3InputShare {
+            input_share: Share::Leader(leader_input_share),
+            proof_share: Share::Leader(leader_proof_share),
+            joint_rand_seed_hint: leader_joint_rand_seed_hint,
+            blind: leader_blind,
         });
-    }
 
-    Ok(out)
+        for helper in helper_shares.into_iter() {
+            out.push(Prio3InputShare {
+                input_share: Share::Helper(helper.input_share),
+                proof_share: Share::Helper(helper.proof_share),
+                joint_rand_seed_hint: helper.joint_rand_seed_hint,
+                blind: helper.blind,
+            });
+        }
+
+        Ok(out)
+    }
 }
 
-/// The verification parameter used by each aggregator to evaluate the VDAF.
-#[derive(Clone, Debug)]
-pub struct VerifyParam<V: Type> {
-    /// Type.
-    pub typ: V,
+/// State of each aggregator during the Prepare process.
+#[allow(missing_docs)]
+pub enum Prio3PrepareStep<F> {
+    /// Ready to send the verifier message.
+    Ready {
+        output_share: Vec<F>,
+        joint_rand_seed: Key,
+        verifier_msg: Prio3VerifierMessage<F>,
+    },
+    /// Waiting for the set of verifier messages.
+    Waiting {
+        output_share: Vec<F>,
+        joint_rand_seed: Key,
+    },
+}
 
-    /// Key used to derive the query randomness from the nonce.
-    pub query_rand_init: Key,
+impl<T: Type, A> Aggregator for Prio3<T, A> {
+    type PrepareStep = Prio3PrepareStep<T::Field>;
+    type PrepareMessage = Prio3VerifierMessage<T::Field>;
 
-    /// The identity of the aggregator.
-    pub aggregator_id: u8,
+    /// Begins the Prep process with the other aggregators. The result of this process is
+    /// the aggregator's output share.
+    //
+    // TODO(cjpatton) Check that the input share matches `self.suite`.
+    fn prepare_init(
+        &self,
+        verify_param: &Prio3VerifyParam,
+        _agg_param: &(),
+        nonce: &[u8],
+        msg: &Prio3InputShare<T::Field>,
+    ) -> Result<Prio3PrepareStep<T::Field>, VdafError> {
+        let mut deriver = KeyDeriver::from_key(&verify_param.query_rand_init);
+        deriver.update(&[255]);
+        deriver.update(nonce);
+        let query_rand_seed = deriver.finish();
+
+        // Create a reference to the (expanded) input share.
+        let expanded_input_share: Option<Vec<T::Field>> = match msg.input_share {
+            Share::Leader(_) => None,
+            Share::Helper(ref seed) => {
+                let prng = Prng::from_key_stream(KeyStream::from_key(seed));
+                Some(prng.take(self.typ.input_len()).collect())
+            }
+        };
+        let input_share = match msg.input_share {
+            Share::Leader(ref data) => data,
+            Share::Helper(_) => expanded_input_share.as_ref().unwrap(),
+        };
```

Comment:
I'm not crazy about this construction, but I can see the bind you're in (Prng output would get dropped within the `Share::Helper` case so you can't return a reference to it) and I can't think of how to do better just now.

Reply:
Ack. I'm wondering how common this problem is and whether there is a better (i.e., more common) pattern for it. We have to do it twice, so anything we can do to make this a bit more DRY would be great.